File: 1/lab-3-1.cpp
#include <iostream>
#include <iomanip>
#include <sstream>
#include <string>
#include <fstream>
#include <cmath>
using namespace std;

double function (double x);

size_t columns_width (size_t titleWidth, double x0, double x1, int prec)
{
	size_t x0_len = (stringstream {} << fixed << setprecision(prec) << x0).str().size();
	size_t x1_len = (stringstream {} << fixed << setprecision(prec) << x1).str().size();

	return max(titleWidth, max(x0_len, x1_len));
}

int main()
{
	double x0 = 3.12345, x1 = 6, step = 0.1;

	string title_arg = "argument:";
	string title_res = "results:";

	size_t width = columns_width(title_arg.size(), x0, x1, 3) + 1;

	cout << "Table" << endl;
	cout << left << setw(width) << title_arg << title_res << endl;
	cout << fixed << setprecision(3);

	for (; x0 <= x1; x0 += step)
	{
		x0 = (abs(x0) < step / 10) ? 0 : x0;
		cout << left << setw(width) << x0 << function(x0) << endl;
	}	
}

double func1 (double x)
{
	return -(x + 3) / 2;
}

double func2 (double x)
{
	return 2 * cos(2 * x);
}

double func3 (double x)
{
	return (x + 3) / 2;
}

double function (double x)
{
	if (x <= -2)
	{
		return func1 (x);
	}
	else if (-2 < x and x < 2)
	{
		return func2 (x);
	}
	else
	{
		return func3 (x);
	}
}

File: lab-3-1.hs
import Text.Printf
import Data.List

function :: Double -> Double
function x
    | x <= -2 = func1 x
    | -2 < x && x < 2 = func2 x
    | otherwise = func3 x

columnsWidth :: Int -> Double -> Double -> Int -> Int
columnsWidth titleWidth x0 x1 prec = maximum [titleWidth, x0Len, x1Len]
    where
        x0Len = length $ printf "%.*f" prec x0
        x1Len = length $ printf "%.*f" prec x1

main :: IO ()
main = do
    let x0 = 3.12345
        x1 = 6
        step = 0.1
        titleArg = "x:"
        titleRes = "results:"
        width = columnsWidth (length titleArg) x0 x1 3 + 1
    putStrLn "Table"
    putStrLn $ printf "%-*s%s" width titleArg titleRes
    forM_ [x0, x0 + step .. x1] $ \x -> do
        let newX = if abs x < step / 10 then 0 else x
        putStrLn $ printf "%-*.*f%.*f" width 3 newX (function newX)

func1 :: Double -> Double
func1 x = -(x + 3) / 2

func2 :: Double -> Double
func2 x = 2 * cos (2 * x)

func3 :: Double -> Double
func3 x = (x + 3) / 2
